compiler/rustc_parse/src/lexer/mod.rs

```
@@ -55,6 +55,16 @@ pub(crate) fn lex_token_trees<'psess, 'src>(
         start_pos = start_pos + BytePos::from_usize(shebang_len);
     }
 
+    // Skip frontmatter, if present.
+    if let Some(frontmatter_len) = rustc_lexer::strip_frontmatter(src) {
```

Other places we call `strip_shebang` that I have not (yet?) updated
- `src/tools/rust-analyzer/crates/parser/src/lexed_str.rs`
- `compiler/rustc_ast_pretty/src/pprust/state.rs`

This is in part because I'm not too sure what the appropriate way of doing it is.

---

> Other places we call strip_shebang that I have not (**yet**?) updated [emphasis: mine]

Both of these aren't blocking this PR, that's for sure.

> compiler/rustc_ast_pretty/src/pprust/state.rs

Disclaimer: I'm not 100% sure what would happen if you don't strip the frontmatter (e.g., whether can lead to ICEs or only butchered output) but I'm certain that you should strip it because otherwise it would reinterpret the frontmatter as Rust code when trying collecting Rust source code comments (which it tries to "reattach" later on).

I mean you could simply check what happens under `-Zunpretty=normal` given a frontmatter that contain `// …` or under `-Zunpretty=expanded` given "macro invocations" inside the frontmatter.

> src/tools/rust-analyzer/crates/parser/src/lexed_str.rs

Note: I don't speak for the rust-analyzer team. However, I know that they prefer that one doesn't modify rust-analyzer from within this repository ([via](https://github.com/rust-lang/rust/blob/827a0d638dabc9a22c56f9c37a557568f86ac76c/triagebot.toml#L878-L879)). So I recommend you to either open an issue or a PR over at rust-lang/rust-analyzer for frontmatter support.

---

I don't think the way this is currently handled is correct (alluded to by fmease above). If we detect a frontmatter, we should parse it as normal, erroring instead of ignoring it if it fails to parse (we currently do the latter and return `Option`).

This needs to be happen before stabilization, and I think it is worth it to do it in this PR too.

---

@fee1-dead I'm not seeing how that is related to this thread, should you open a new one?  I have questions on your reply but I don't want to distract from whats being covered here (where all we should strip, how, and if its in this PR)

---

> src/tools/rust-analyzer/crates/parser/src/lexed_str.rs

I've made a separate line item on the tracking issue for r-a's implementation

> compiler/rustc_ast_pretty/src/pprust/state.rs

Knowing what to look for in tests, I've gone ahead and add this to this PR, updating the initial test commit and adding a commit specifically for pretty support.